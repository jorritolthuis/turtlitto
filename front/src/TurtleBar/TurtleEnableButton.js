import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { buttonColor } from "../theme.js";

/**
 *
 * Author: S.A. Tanja
 * Author: H.E. van der Laan
 * Author: T.T.P. Franken
 *
 *
 * A button that keeps track whether a turtle is enabled or not.
 * Is reflected to the user in terms in terms of button appearance.
 *
 * Props:
 * - enabled: whether this turtle is enabled or not
 * - onDisable: function to call when disabling this turtle
 * - onEnable: function to call when enabling this turtle
 *
 */
const TurtleEnableButton = props => {
  if (props.enabled) {
    return (
      <Button className={props.className} isActive onClick={props.onDisable}>
        {props.id}
      </Button>
    );
  } else {
    return (
      <Button className={props} onClick={props.onEnable}>
        {props.id}
      </Button>
    );
  }
};

TurtleEnableButton.propTypes = {
  enabled: PropTypes.bool.isRequired,
  onDisable: PropTypes.func.isRequired,
  onEnable: PropTypes.func.isRequired
};

const Button = styled.button`
  border-style: ${props => (props.isActive ? "inset" : "solid")};
  border-color: ${props => (props.isActive ? "none" : buttonColor)};
  background-color: ${props => (props.isActive ? "silver" : buttonColor)};
  width: 16%;
  min-width: 7.5rem;
  height: 10rem;
  font-size: 4rem;
  flex: 1;
  margin: 0.1rem;
  &:active {
    background: "silver";
  }
`;

export default TurtleEnableButton;
