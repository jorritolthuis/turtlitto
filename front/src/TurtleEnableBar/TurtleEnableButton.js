import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

/**
 * A button that keeps track whether the turtle is enabled.
 * It is reflected to the user in terms in terms of button appearance.
 * Author: S.A. Tanja
 * Author: H.E. van der Laan
 * Author: T.T.P. Franken
 *
 * Props:
 *  - enabled: a boolean flag specifying whether is enabled or not
 *  - onTurtleEnableChange: function to call when the turtle enable button is pressed
 *  - id: identifier of a turtle
 *  - className: gives the classname for css
 */
const TurtleEnableButton = props => {
  const { enabled, id, className, onTurtleEnableChange } = props;

  return (
    <Button
      className={className}
      isActive={enabled}
      onClick={onTurtleEnableChange}
    >
      {id}
    </Button>
  );
};

TurtleEnableButton.propTypes = {
  enabled: PropTypes.bool.isRequired,
  onTurtleEnableChange: PropTypes.func.isRequired
};

const Button = styled.button`
  border-style: ${props =>
    props.isActive
      ? props.theme.buttonBorderStyleActive
      : props.theme.buttonBorderStyle};
  border-color: ${props =>
    props.isActive ? "none" : props.theme.buttonBorder};
  background-color: ${props =>
    props.isActive ? props.theme.buttonActive : props.theme.button};
  font-size: 4rem;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 2rem;
  min-width: 7rem;
  flex: 1;
`;
Button.displayName = "TurtleEnableButton";

export default TurtleEnableButton;
