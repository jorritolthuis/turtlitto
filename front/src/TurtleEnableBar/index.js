import React from "react";
import TurtleEnableButton from "./TurtleEnableButton";
import styled from "styled-components";
import PropTypes from "prop-types";

/**
 * Simple bar with all the turtles
 * Author: S.A. Tanja
 * Author: H.E. van der Laan
 * Author: T.T.P. Franken
 * Author: B. Afonins
 *
 * Props:
 * - turtles: a list of turtles to be displayed in the bar
 *   - id: an identificator of the turtle
 *   - enabled: a boolean flag specifying whether the turtle is enabled
 * - onDisable: function to call when disabling this turtle
 * - onEnable: function to call when enabling this turtle
 */

const TurtleEnableBar = props => {
  return (
    <Bar className={props.className}>
      {props.turtles.map((turtle, position) => {
        return (
          <TurtleEnableButton
            key={turtle.id}
            enabled={turtle.enabled}
            id={position + 1}
            onEnable={() => props.onEnable(position)}
            onDisable={() => props.onDisable(position)}
          />
        );
      })}
    </Bar>
  );
};

TurtleEnableBar.propTypes = {
  turtles: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      enabled: PropTypes.bool.isRequired
    })
  ).isRequired,
  onDisable: PropTypes.func.isRequired,
  onEnable: PropTypes.func.isRequired
};

const Bar = styled.div`
  display: flex;
  height: 15%;
  width: 100%;
  justify-content: space-between;
  overflow-x: auto;
  align-content: space-between;
  border-style: solid;
  border-width: 0rem 0rem 0.2rem 0rem;
  margin-bottom: 0.2rem;
  background: ${props => props.theme.turtleEnableBar};
  position: sticky;
  z-index: 1;
`;

export default TurtleEnableBar;
