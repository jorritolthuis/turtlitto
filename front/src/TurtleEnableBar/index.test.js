import React from "react";
import { shallow, mount } from "enzyme";
import TurtleEnableBar from ".";
import TurtleEnableButton from "./TurtleEnableButton";

describe("TurtleEnableBar", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(
      <TurtleEnableBar turtles={[]} onTurtleEnableChange={() => {}} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should render all turtles as <TurtleEnableButton /> components", () => {
    const turtles = [
      {
        id: 0,
        enabled: false
      },
      {
        id: 1,
        enabled: false
      },
      {
        id: 2,
        enabled: false
      }
    ];
    const wrapper = mount(
      <TurtleEnableBar turtles={turtles} onTurtleEnableChange={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(TurtleEnableButton).length).toBe(turtles.length);
  });
});
