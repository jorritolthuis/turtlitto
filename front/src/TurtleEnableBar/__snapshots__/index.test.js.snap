// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurtleEnableBar should match snapshot 1`] = `<styled.div />`;

exports[`TurtleEnableBar should render all turtles as <TurtleEnableButton /> components 1`] = `
<TurtleEnableBar
  onTurtleEnableChange={[Function]}
  turtles={
    Array [
      Object {
        "enabled": false,
        "id": 0,
      },
      Object {
        "enabled": false,
        "id": 1,
      },
      Object {
        "enabled": false,
        "id": 2,
      },
    ]
  }
>
  <styled.div>
    <div
      className="sc-bwzfXH byJVkT"
    >
      <TurtleEnableButton
        enabled={false}
        id={1}
        key="0"
        onTurtleEnableChange={[Function]}
      >
        <styled.button
          isActive={false}
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa fIMHSN"
            onClick={[Function]}
          >
            1
          </button>
        </styled.button>
      </TurtleEnableButton>
      <TurtleEnableButton
        enabled={false}
        id={2}
        key="1"
        onTurtleEnableChange={[Function]}
      >
        <styled.button
          isActive={false}
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa fIMHSN"
            onClick={[Function]}
          >
            2
          </button>
        </styled.button>
      </TurtleEnableButton>
      <TurtleEnableButton
        enabled={false}
        id={3}
        key="2"
        onTurtleEnableChange={[Function]}
      >
        <styled.button
          isActive={false}
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa fIMHSN"
            onClick={[Function]}
          >
            3
          </button>
        </styled.button>
      </TurtleEnableButton>
    </div>
  </styled.div>
</TurtleEnableBar>
`;
